model/infinispan/src/main/java/org/keycloak/keys/infinispan/InfinispanPublicKeyStorageProvider.java

```diff
@@ -139,28 +144,97 @@ public KeyWrapper getPublicKey(String modelKey, String kid, String algorithm, Pu
             }
         }
 
+        PublicKeysEntry updatedEntry = reloadKeys(modelKey, entry, currentTime, loader);
+        entry = updatedEntry == null? entry : updatedEntry;
+        KeyWrapper publicKey = entry == null? null : entry.getCurrentKeys().getKeyByKidAndAlg(kid, algorithm);
+        if (publicKey != null) {
+            // return a copy of the key to not modify the cached one
+            return publicKey.cloneKey();
+        }
+
+        List<String> availableKids = entry == null? Collections.emptyList() : entry.getCurrentKeys().getKids();
+        log.warnf("PublicKey wasn't found in the storage. Requested kid: '%s' . Available kids: '%s'", kid, availableKids);
+
+        return null;
+    }
+
+    /**
+     * If the key is found in the cache that is returned straight away. If not in cache,
+     * the keys are reloaded if allowed by the minTimeBetweenRequests and key
+     * is searched again.
+     *
+     * @param modelKey The model key
+     * @param predicate The predicate to search the key
+     * @param loader The loader to reload keys
+     * @return The key or null
+     */
+    @Override
+    public KeyWrapper getPublicKey(String modelKey, Predicate<KeyWrapper> predicate, PublicKeyLoader loader) {
+        PublicKeysEntry entry = keys.get(modelKey);
+        if (entry != null) {
+            // if in cache just try to return if found
+            KeyWrapper key = entry.getCurrentKeys().getKeyByPredicate(predicate);
+            if (key != null) {
+                return key;
```

Comment:
Not sure, but don't you want to clone here too as you are doing in line 151 above?

Reply:
I missed this one, yes, you are right, Done now!